cpp/tests/spatial/linestring_distance_test.cpp

```diff
@@ -0,0 +1,399 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuspatial/distances/linestring_distance.hpp>
+#include <cuspatial/error.hpp>
+
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/column_utilities.hpp>
+#include <cudf_test/column_wrapper.hpp>
+#include <cudf_test/type_lists.hpp>
+
+namespace cuspatial {
+namespace test {
+
+using namespace cudf;
+using namespace cudf::test;
+
+template <typename T>
+using wrapper = fixed_width_column_wrapper<T>;
+
+template <typename T>
+struct PairwiseLinestringDistanceTest : public BaseFixture {
+};
+
+// float and double are logically the same but would require separate tests due to precision.
+using TestTypes = FloatingPointTypes;
+TYPED_TEST_CASE(PairwiseLinestringDistanceTest, TestTypes);
+
+constexpr cudf::test::debug_output_level verbosity{cudf::test::debug_output_level::ALL_ERRORS};
+
```

Comment:
One degenerate test I would like to see is one that exercises this:
```
auto r_denom = cross(ab, cd);
  if (r_denom == 0) {  // should we be using an approximate equals or < epsilon to avoid division by a very small number?
    // Segments parallel or collinear
    return segment_distance_no_intersect_or_collinear(A, B, C, D);
  }
  auto r_numer = cross(ac, cd); // note I changed CA in your code to AC here for consistency
  auto r = r_numer / r_denom;
```

Specifically, the case where `r_denom` is very small but non-zero (possibly denormalized).  So the test could be using two segments that are very nearly but not quite coincident and non-parallel.  If the denominator ends up very small, then you could end up with a very large r. This should be OK, but it's good to test to ensure it is robust.

Reply:
Test added!